File: src/lib/schemas/syllabus.ts
import * as z from "zod"

import { IPublishStatus } from "@/types/syllabus.interface"

export const SyllabusNewSchema = z
  .object({
    topicCode: z
      .string()
      .min(3, { message: "Syllabus code must be at least 3 characters" }),
    topicName: z.string().min(1, { message: "Syllabus name is required" }),
    technicalRequirement: z
      .string()
      .min(1, { message: "Technical requirement is required" }),
    courseObjective: z
      .string()
      .min(1, { message: "Course objective is required" }),
    trainingPrinciple: z.string().min(1, { message: "Principle is required" }),
    assessment: z.object({
      assessmentID: z.string(),
      quizCount: z.coerce
        .number()
        .positive()
        .min(0, { message: "Quiz count is required" }),
      quizPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Quiz percent is required" }),
      assignmentCount: z.coerce
        .number()
        .positive()
        .min(0, { message: "Assignment count is required" }),
      assignmentPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Assignment percent is required" }),
      finalTheoryPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Final theory percent is required" }),
      finalPracticePercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Final practice percent is required" })
    })
  })
  .refine(
    (data) => {
      const {
        assignmentPercent,
        quizPercent,
        finalTheoryPercent,
        finalPracticePercent
      } = data.assessment

      return (
        assignmentPercent +
          quizPercent +
          finalTheoryPercent +
          finalPracticePercent ===
        100
      )
    },
    {
      message: "The sum of all percentages must be equal to 100",
      path: ["assessment"]
    }
  )

export const SyllabusSchema = z
  .object({
    topicCode: z
      .string()
      .min(3, { message: "Syllabus code must be at least 3 characters" }),
    topicName: z.string().min(1, { message: "Syllabus name is required" }),
    version: z.coerce.number(),
    technicalGroup: z.string().min(1, { message: "Group is required" }),
    trainingAudience: z.string().min(1, { message: "Audience is required" }),
    topicOutline: z.string().min(1, { message: "Outline is required" }),
    trainingMaterials: z.string().min(1, { message: "Materials is required" }),
    createdBy: z.string().min(1, { message: "Created by is required" }),
    createdDate: z
      .string()
      .min(1, { message: "Created date is required" })
      .or(z.date()),
    modifiedBy: z.string().min(1, { message: "Modified by is required" }),
    modifiedDate: z
      .string()
      .min(1, { message: "Modified date is required" })
      .or(z.date()),
    publishStatus: z.nativeEnum(IPublishStatus),
    technicalRequirement: z
      .string()
      .min(1, { message: "Technical requirement is required" }),
    courseObjective: z
      .string()
      .min(1, { message: "Course objective is required" }),
    trainingPrinciple: z.string().min(1, { message: "Principle is required" }),
    assessment: z.object({
      assessmentID: z.string(),
      quizCount: z.coerce
        .number()
        .positive()
        .min(0, { message: "Quiz count is required" }),
      quizPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Quiz percent is required" }),
      assignmentCount: z.coerce
        .number()
        .positive()
        .min(0, { message: "Assignment count is required" }),
      assignmentPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Assignment percent is required" }),
      finalTheoryPercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Final theory percent is required" }),
      finalPracticePercent: z.coerce
        .number()
        .positive()
        .min(0, { message: "Final practice percent is required" })
    }),
    trainingUnits: z.array(z.any())
  })
  .refine(
    (data) => {
      const {
        assignmentPercent,
        quizPercent,
        finalTheoryPercent,
        finalPracticePercent
      } = data.assessment

      return (
        assignmentPercent +
          quizPercent +
          finalTheoryPercent +
          finalPracticePercent ===
        100
      )
    },
    {
      message: "The sum of all percentages must be equal to 100",
      path: ["assessment"]
    }
  )

File: fams_08042024_2307/fams-main/DataLayer/Entities/Syllabus.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DataLayer.Entities
{
    public class Syllabus
    {
        [Key]
        [Required(ErrorMessage = "Topic Code is not null"), NotNull]
        [StringLength(10, ErrorMessage = "Topic Code is not longer than 10")]
        public string? TopicCode { get; set; }

        [Required(ErrorMessage = "Topic Name is not null"), NotNull]
        [StringLength(100, ErrorMessage = "Topic Name is not longer than 100")]
        public string? TopicName { get; set; }

        [Required(ErrorMessage = "Technical Group is not null"), NotNull]
        [StringLength(100, ErrorMessage = "Technical Group is not longer than 100")]
        public string? TechnicalGroup { get; set; }

        [StringLength(int.MaxValue)]
        [Required(ErrorMessage = "Technical Requirement is not null"), NotNull]
        public string TechnicalRequirement { get; set; }

        [StringLength(int.MaxValue)]
        [Required(ErrorMessage = "Course Objective is not null"), NotNull]
        public string CourseObjective { get; set; }

        [Required(ErrorMessage = "Version is not null"),]
        public int Version { get; set; }

        [Required(ErrorMessage = "Training Audience is not null"), NotNull]
        [StringLength(100, ErrorMessage = "Training Audience is not longer than 100")]
        public string? TrainingAudience { get; set; }

        [Required(ErrorMessage = "Topic Outline is not null")]
        public string? TopicOutline { get; set; }

        public string? TrainingMaterials { get; set; }

        public string? TrainingPrinciple { get; set; }

        [Required(ErrorMessage = "Not null"), NotNull]
        public Priorities Priority { get; set; }

        [Required(ErrorMessage = "Not null"), NotNull]
        public PulishStatuses PulishStatus { get; set; }

        public string? CreatedBy { get; set; }

        public DateTime CreatedDate { get; set; }

        public string? ModifiedBy { get; set; }
        public DateTime ModifiedDate { get; set; }

        public Guid UserId { get; set; }

        public string? AssessmentID { get; set; }
        public virtual User? User { get; set; }
        public virtual Assessment? Assessment { get; set; }
        public virtual ICollection<TrainingProgramSyllabus>? TrainingProgramSyllabuses { get; set; }
        public virtual ICollection<SyllabusObjective>? SyllabusObjectives { get; set; }
        public virtual ICollection<TrainingUnit>? TrainingUnits { get; set; }

        public enum Priorities
        {
            Low,
            Mid,
            High
        }

        public enum PulishStatuses
        {
            Denied,
            Editing,
            Pending,
            Published
        }

    }
}
